desktop/desktopentry/desktopentry.go

```diff
@@ -0,0 +1,236 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package desktopentry
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/strutil"
+)
+
+// DesktopEntry represents a freedesktop.org desktop entry file.
+//
+// The various fields are as defined in the specification:
+// https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html
+type DesktopEntry struct {
+	Filename string
+	Name     string
+	Icon     string
+	Exec     string
+
+	Hidden                bool
+	OnlyShowIn            []string
+	NotShownIn            []string
+	GnomeAutostartEnabled bool
+
+	Actions map[string]*Action
+}
+
+// Action represents an application action defined in a desktop entry file.
+type Action struct {
+	Name string
+	Icon string
+	Exec string
+}
+
+type groupState int
+
+const (
+	unknownGroup groupState = iota
+	desktopEntryGroup
+	desktopActionGroup
+)
+
+func splitList(value string) []string {
+	return strings.FieldsFunc(value, func(r rune) bool { return r == ';' })
+}
+
+// Read parses a desktop entry file.
+func Read(filename string) (*DesktopEntry, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	return parse(filename, f)
+}
+
+func parse(filename string, r io.Reader) (*DesktopEntry, error) {
+	de := &DesktopEntry{
+		Filename:              filename,
+		GnomeAutostartEnabled: true,
+	}
+	var (
+		currentGroup          = unknownGroup
+		seenDesktopEntryGroup = false
+		actions               []string
+		currentAction         *Action
+	)
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		// Ignore empty lines and comments
+		if line == "" || strings.HasPrefix(line, "#") {
+			continue
+		}
+		// Start of a new group
+		if strings.HasPrefix(line, "[") {
+			if line == "[Desktop Entry]" {
+				if seenDesktopEntryGroup {
+					return nil, fmt.Errorf("desktop file %q has multiple [Desktop Entry] groups", filename)
+				}
+				seenDesktopEntryGroup = true
+				currentGroup = desktopEntryGroup
+			} else if strings.HasPrefix(line, "[Desktop Action ") && strings.HasSuffix(line, "]") {
+				action := line[len("[Desktop Action ") : len(line)-1]
+				if !strutil.ListContains(actions, action) {
+					return nil, fmt.Errorf("desktop file %q contains unknown action %q", filename, action)
+				}
+				if de.Actions[action] != nil {
+					return nil, fmt.Errorf("desktop file %q has multiple %q groups", filename, line)
+				}
+				currentGroup = desktopActionGroup
+				if de.Actions == nil {
+					de.Actions = make(map[string]*Action, len(actions))
+				}
+				currentAction = &Action{}
+				de.Actions[action] = currentAction
+			} else {
+				// Ignore other groups
+				currentGroup = unknownGroup
+			}
+			continue
+		}
+
+		split := strings.SplitN(line, "=", 2)
+		if len(split) != 2 {
+			return nil, fmt.Errorf("desktop file %q badly formed", filename)
```

Comment:
(nitpick) maybe `return nil, fmt.Errorf("desktop file %q badly formed in line %q", filename, line)`